### ISSUE TITLE
Does FFI support iOS static library？

### ISSUE THREAD

**Issue by JackPanda8** [dart-lang/sdk#44328]
as the official doc :
**Symbols from a statically linked library can be loaded using DynamicLibrary.executable or DynamicLibrary.process.**

but i got the error: Invalid argument(s): Failed to lookup symbol (dlsym(RTLD_DEFAULT, pause_all_task): symbol not found)

-------------here is my code--------------------------------
```dart
import 'dart:ffi' as ffi;
import 'dart:io';
import 'package:ffi/ffi.dart';

// C function signatures
typedef pause_all_task_t = ffi.Void Function();
typedef get_cache_instance = ffi.Pointer<ffi.Void> Function();

// Dart function signatures
typedef PauseALLTask = void Function();
typedef GetCacheInstance = ffi.Pointer<ffi.Void> Function();

// Getting a library that holds needed symbols
ffi.DynamicLibrary _lib = Platform.isAndroid
		? ffi.DynamicLibrary.open('libflash-downloader-lib.so')
		: ffi.DynamicLibrary.process();

// Looking for the functions
final PauseALLTask pauseAllTaskF = _lib
		.lookup<ffi.NativeFunction<pause_all_task_t>>('pause_all_task')
		.asFunction();
final GetCacheInstance getCacheInstanceF = _lib
		.lookup<ffi.NativeFunction<get_cache_instance>>('get_cache_instance')
		.asFunction();
```

```c++
#include "dart_ffi_adapter.h"
//#include "YKCacheCenter.h"
extern "C" {
__attribute__((visibility("default"))) __attribute__((used))
void* get_cache_instance(void) {
    printf("my_download:调用GetCacheInstance构造单例");
    void* p;
    return p;
}
__attribute__((visibility("default"))) __attribute__((used))
void pause_all_task(void) {
    //    yk_cache::IYKCache *cache = yk_cache::GetCacheInstance();
    //    cache->pauseAllTask();
    printf("my_download:调用pause_all_task暂停所有下载中的任务");
}
__attribute__((visibility("default"))) __attribute__((used))
void ff_pause_all_task(){
    printf("my_download:call ff_pause_all_task");
}
}
```

**Comment by dcharkes** [comment]
Hi @JackPanda8,

Please take a look at https://github.com/flutter/flutter/issues/62666#issuecomment-668377567 

And please double check that Xcode statically links the C code:

> On iOS, you need to tell Xcode to statically link the file:
>
> In Xcode, open Runner.xcworkspace.
> Add the C/C++/Objective-C/Swift source files to the Xcode project.

https://flutter.dev/docs/development/platform-integration/c-interop#step-2-add-cc-sources

If the problem still persists, please provide a minimal reproduction.

**Comment by bitsydarel** [comment]
Header and universal static library are added

<img width="256" alt="Capture d’écran 2021-05-20 à 16 46 15" src="https://user-images.githubusercontent.com/17922775/118989769-f7443580-b98a-11eb-9941-eb0f51ecf04f.png">

Strip style set to non-global

<img width="846" alt="Capture d’écran 2021-05-20 à 16 47 29" src="https://user-images.githubusercontent.com/17922775/118989895-13e06d80-b98b-11eb-8d4c-8c989d19517e.png">

Even added the attribute about visibility

<img width="846" alt="Capture d’écran 2021-05-20 à 16 48 42" src="https://user-images.githubusercontent.com/17922775/118990134-4ab68380-b98b-11eb-9d23-d96cf8006365.png">

It's not working even debug build on simulator

<img width="1373" alt="Capture d’écran 2021-05-20 à 16 51 06" src="https://user-images.githubusercontent.com/17922775/118990462-95d09680-b98b-11eb-95eb-e4ac262f1531.png">

Even the iOS swift app delegate can see and run it without problem

<img width="1142" alt="Capture d’écran 2021-05-20 à 17 07 34" src="https://user-images.githubusercontent.com/17922775/118993194-fb258700-b98d-11eb-94ed-50ca3bcded27.png">

the universal static library support all the cpu

<img width="855" alt="Capture d’écran 2021-05-20 à 17 09 18" src="https://user-images.githubusercontent.com/17922775/118993395-27d99e80-b98e-11eb-8852-f959183bfc93.png">

even create a test c file 
<img width="242" alt="Capture d’écran 2021-05-20 à 17 29 58" src="https://user-images.githubusercontent.com/17922775/118997023-0201c900-b991-11eb-81eb-24da803fc484.png">

added pseudo code that internally call the library function

<img width="752" alt="Capture d’écran 2021-05-20 à 17 31 21" src="https://user-images.githubusercontent.com/17922775/118997249-337a9480-b991-11eb-80f7-425882840128.png">

<img width="671" alt="Capture d’écran 2021-05-20 à 17 31 44" src="https://user-images.githubusercontent.com/17922775/118997335-41301a00-b991-11eb-896a-162672c176ec.png">

<img width="722" alt="Capture d’écran 2021-05-20 à 17 35 53" src="https://user-images.githubusercontent.com/17922775/118998095-d501e600-b991-11eb-90d0-15dcc31ee8da.png">

The number "3" is successfully printed in the console so the FPDF_DestroyLibrary symbol is definetly not stripped.
<img width="400" alt="Capture d’écran 2021-05-20 à 17 35 15" src="https://user-images.githubusercontent.com/17922775/118997997-c3b8d980-b991-11eb-8ad1-c769b55565f9.png">

But still dart can't see it...

<img width="835" alt="Capture d’écran 2021-05-20 à 17 36 37" src="https://user-images.githubusercontent.com/17922775/118998223-efd45a80-b991-11eb-8cbf-fc87a0692d92.png">

**Comment by bitsydarel** [comment]
Here are the necessary files need to to replicate
https://drive.google.com/file/d/1tpI_StOh7ffPHjVapJJ34dMg2tJka45E/view?usp=sharing

**Comment by bitsydarel** [comment]
Any update or comment?

**Comment by dcharkes** [comment]
Hi @bitsydarel, I have not yet had time to look into this.

The last time I tried to do static linking for iOS and MacOS I ran into similar problems. I ended up using dynamic libraries instead.

Does using dynamic libraries work for you in the mean time? (Until I have time to dig into this.)

**Comment by bitsydarel** [comment]
@dcharkes not really as those libraries are provided to us, and I have not seen usecase of creating fat library as dynamic libraries (for simulator and device).

**Comment by bitsydarel** [comment]
@dcharkes Does it have to do with how dart lookup for library ? Is there anyway I could help to speed up the résolution of this issue ? As it’s one of our blocker in an ongoing project.

**Comment by dcharkes** [comment]
> Here are the necessary files need to to replicate
> https://drive.google.com/file/d/1tpI_StOh7ffPHjVapJJ34dMg2tJka45E/view?usp=sharing

I don't have access to this.

Could you upload it is a GitHub repository instead?



> The last time I tried to do static linking for iOS and MacOS I ran into similar problems. I ended up using dynamic libraries instead.

I have succeeded in looking up symbols of a statically linked library by creating the podspec of the framework as the following.

`flutter create --platforms=android,ios,macos,windows,linux --template=plugin mylib_staticlib`

Modify `mylib_staticlib/macos/mylib_staticlib.podspec` (for Flutter desktop MacOS):

```podspec
Pod::Spec.new do |s|
  s.name             = 'mylib_staticlib'
  s.version          = '0.0.1'
  s.summary          = 'A new flutter plugin project.'
  s.description      = <<-DESC
A new flutter plugin project.
                       DESC
  s.homepage         = 'http://example.com'
  s.license          = { :file => '../LICENSE' }
  s.author           = { 'Your Company' => 'email@example.com' }
  s.source           = { :path => '.' }
  s.source_files     = 'Classes/**/*'
  s.dependency 'FlutterMacOS'

  s.platform = :osx, '10.11'
  s.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES' }
  s.vendored_libraries = 'Frameworks/libmylib_staticlib.a'
  s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-force_load $(PODS_TARGET_SRCROOT)/Frameworks/libmylib_staticlib.a" }
  s.swift_version = '5.0'
end
```

And copy the static library in to `mylib_staticlib/macos/Frameworks`.

For iOS, a similar approach.

`mylib_staticlib/ios/mylib_staticlib.podspec`:

```podspec
Pod::Spec.new do |s|
  s.name             = 'mylib_staticlib'
  s.version          = '0.0.1'
  s.summary          = 'A new flutter plugin project.'
  s.description      = <<-DESC
A new flutter plugin project.
                       DESC
  s.homepage         = 'http://example.com'
  s.license          = { :file => '../LICENSE' }
  s.author           = { 'Your Company' => 'email@example.com' }
  s.source           = { :path => '.' }
  s.source_files = 'Classes/**/*'
  s.dependency 'Flutter'
  s.platform = :ios, '8.0'

  # Flutter.framework does not contain a i386 slice.
  s.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES', 'EXCLUDED_ARCHS[sdk=iphonesimulator*]' => 'i386' }
  s.vendored_libraries = 'Frameworks/libmylib_staticlib.a'
  s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-force_load $(PODS_TARGET_SRCROOT)/Frameworks/libmylib_staticlib.a" }
  s.swift_version = '5.0'
end
```

And the fat static library (arm64 for device and x64 for simulator) in `mylib_staticlib/ios/Frameworks/libmylib_staticlib.a`.

So the 2 important lines in the podspec:

```podspec
  s.vendored_libraries = 'Frameworks/libmylib_staticlib.a'
  s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-force_load $(PODS_TARGET_SRCROOT)/Frameworks/libmylib_staticlib.a" }
```

The first statically links the static library, and the second line is the correct invocation to tell the linker to not strip the symbols.

**Comment by bitsydarel** [comment]
Here's the repository with the necessary files
https://github.com/bitsydarel/dart_ffi_static_link_issue

**Comment by bitsydarel** [comment]
@dcharkes  Would be nice to have this test on iOS and on template=app

**Comment by dcharkes** [comment]
To troubleshoot your problem, we need a complete flutter project, not just the static libraries and header files.

Create a new one with `flutter create --platforms=android,ios,macos,windows,linux --template=plugin <insert-name>`, copy the static library in to the flutter plugin, modify the podspec, and try to lookup the symbol in a simple flutter app causing the error.

**Comment by dcharkes** [comment]
> @dcharkes Would be nice to have this test on iOS and on template=app

You are always going to need a `template=plugin` approach here, both for static libraries, dynamic libraries and source code.

**Comment by bitsydarel** [comment]
@dcharkes So basically meaning adding it in an application directly won't do ? is there a reason for the why the app can't create those build configuration and if possible what are those build configuration ?

**Comment by dcharkes** [comment]
One could possibly. I am not familiar with the how the Flutter app/plugin builds are setup to tell you how. A good starting point for reverse engineering that would be to take a look at what files are generated when doing `flutter create --template=plugin`. And otherwise ask a question on the Flutter issue tracker.

Any specific reason why you don't want to have a Flutter app + Flutter plugin right next to it in the same repository? And then have a `../myplugin` dependency in your `pubspec.yaml`?

**Comment by bitsydarel** [comment]
@dcharkes  project pushed, also added the lookup functions and on the podspec file also added the required setup as suggested.
But still not working on iOS

https://github.com/bitsydarel/dart_ffi_static_link_issue

**Comment by bitsydarel** [comment]
`[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: Invalid argument(s): Failed to lookup symbol (dlsym(RTLD_DEFAULT, FPDF_DestroyLibrary): symbol not found)
#0      DynamicLibrary.lookup (dart:ffi-patch/ffi_dynamic_library_patch.dart:31:29)
#1      new PdfRenderer.load (package:dart_ffii_static_link_issue/src/pdf_renderer.dart:454:12)
#2      DartFfiiStaticLinkIssue.checkPDFiumLink (package:dart_ffii_static_link_issue/dart_ffii_static_link_issue.dart:20:46)
#3      DartFfiiStaticLinkIssue.platformVersion (package:dart_ffii_static_link_issue/dart_ffii_static_link_issue.dart:14:5)
#4      _MyAppState.initPlatformState (package:dart_ffii_static_link_issue_example/main.dart:32:41)
#5      _MyAppState.initState (package:dart_ffii_static_link_issue_example/main.dart:22:5)
#6      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4711:57)
#7      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4548:5)
#8      Element.inflateWidget (package:flutter/src/widgets/fra<…>`

**Comment by dcharkes** [comment]
Your reproduction doesn't work here. The static library is missing symbols. Did you mean to include a static library with the standard c/c++ libs as well?

```
Xcode's output:
↳
    Undefined symbols for architecture arm64:
      "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::setbuf(char*, long)", referenced from:
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(cpdf_streamparser.o)
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(cfdf_document.o)
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(cpdf_syntax_parser.o)
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(fpdf_parser_decode.o)
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(cpdf_generateap.o)
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(cfx_fontmapper.o)
          vtable for std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> > in libpdfium.a(cpdfsdk_appstream.o)

...
```

[full log](https://gist.github.com/dcharkes/199d159586c88e90b7b2dfeb93836a7c)

**Comment by bitsydarel** [comment]
@dcharkes  if you check the android pdfium folder, you will see other libraries that the library would need

**Comment by dcharkes** [comment]
> @dcharkes if you check the android pdfium folder, you will see other libraries that the library would need

https://github.com/bitsydarel/dart_ffi_static_link_issue/tree/main/pdfium/android

Those are all shared libraries, not static libraries, and not compiled in a way MacOS/iOS can understand them.

Please provide a working reproduction.

**Comment by paulo-coutinho** [comment]
Hi @bitsydarel,

This is simple. Add on "OTHER_LDFLAGS" the link to "c++" as i showed before in xcode sample.

File: dart_ffii_static_link_issue.podspec

```
s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-force_load $(PODS_TARGET_SRCROOT)/Frameworks/libpdfium.a -lc++" }
```

**Comment by bitsydarel** [comment]
Hi @paulo-coutinho compilation now's working but it's still can't find the symbol 
`[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: Invalid argument(s): Failed to lookup symbol (dlsym(RTLD_DEFAULT, FPDF_DestroyLibrary): symbol not found)
#0      DynamicLibrary.lookup (dart:ffi-patch/ffi_dynamic_library_patch.dart:31:29)
#1      new PdfRenderer.load (package:dart_ffii_static_link_issue/src/pdf_renderer.dart:454:12)
#2      DartFfiiStaticLinkIssue.checkPDFiumLink (package:dart_ffii_static_link_issue/dart_ffii_static_link_issue.dart:20:46)
#3      DartFfiiStaticLinkIssue.platformVersion (package:dart_ffii_static_link_issue/dart_ffii_static_link_issue.dart:14:5)
#4      _MyAppState.initPlatformState (package:dart_ffii_static_link_issue_example/main.dart:32:41)
#5      _MyAppState.initState (package:dart_ffii_static_link_issue_example/main.dart:22:5)
#6      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4711:57)
#7      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4548:5)
#8      Element.inflateWidget (package:flutter/src/widgets/fra<…>
`

@dcharkes Now the sample is reproducible, i pushed the changes so you could check if it's dart that can't load the symbols.

**Comment by paulo-coutinho** [comment]
How do you testing it?

**Comment by bitsydarel** [comment]
When you run the app if you check the console you should see logs about about the running app.

**Comment by paulo-coutinho** [comment]
Hi,

Im investigating the generated binary from the Runner app and Xcode don't link with it:

```
nm /Users/paulo/Library/Developer/Xcode/DerivedData/PDFiumTest-floybqwuisyrxufsdrwjfipgjiwy/Build/Products/Debug-iphoneos/PDFiumTest.app | grep FPDF
[nothing]
```

And i make the same thing with that my sample using the same library and it is linked correctly:
```
nm /Users/paulo/Library/Developer/Xcode/DerivedData/PDFiumTest-floybqwuisyrxufsdrwjfipgjiwy/Build/Products/Debug-iphoneos/PDFiumTest.app/PDFiumTest | grep FPDF
0000000100069874 t _FPDFPage_GetCropBox
000000010006978c t _FPDFPage_GetMediaBox
0000000100069318 t _FPDFText_CountChars
0000000100069550 t _FPDFText_CountRects
00000001000695cc t _FPDFText_GetBoundedText
0000000100069358 t _FPDFText_GetCharBox
00000001000693f8 t _FPDFText_GetCharIndexAtPos
0000000100069328 t _FPDFText_GetFontSize
000000010006955c t _FPDFText_GetRect
0000000100069440 t _FPDFText_GetText
0000000100069278 t _FPDFText_LoadPage
000000010006a5b8 t _FPDF_CloseDocument
000000010006a4e0 t _FPDF_ClosePage
000000010006a45c t _FPDF_GetPageBoundingBox
000000010006a2d8 t _FPDF_GetPageCount
000000010006a14c t _FPDF_InitLibraryWithConfig
000000010006a1c4 t _FPDF_LoadDocument
000000010006a310 t _FPDF_LoadPage
000000010006898c t __Z20CPDFPageFromFPDFPageP13fpdf_page_t__
0000000100068980 t __Z20FPDFPageFromIPDFPageP9IPDF_Page
000000010006897c t __Z20IPDFPageFromFPDFPageP13fpdf_page_t__
0000000100068984 t __Z28CPDFDocumentFromFPDFDocumentP17fpdf_document_t__
0000000100068988 t __Z28FPDFDocumentFromCPDFDocumentP13CPDF_Document
```

The library is not the problem, but the link process.

**Comment by paulo-coutinho** [comment]
You can use any library name in podspec that it don't will generate error and will open the app:

```
s.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES', 'EXCLUDED_ARCHS[sdk=iphonesimulator*]' => 'i386' }
s.vendored_libraries = 'libpdfiumXYZ.a'
s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-lc++ -lz" }
```

So you problem is masked.

**Comment by bitsydarel** [comment]
If your sample were using podfile, how would you link the library ?

**Comment by paulo-coutinho** [comment]
It is easy. I made it work here:

https://github.com/paulo-coutinho/pdfium-test/pull/1

You need:
```
cd ios
pod install
open PDFiumTest.xcworkspace
```

**Comment by bitsydarel** [comment]
@dcharkes any update because the suggested changes by @paulo-coutinho does actually make the code compile and also make it available in swift or objective c code.

But dart can't see it...

Latest configuration:
`s.source_files     = 'Classes/**/*'
  s.dependency 'Flutter'
  s.platform         = :ios, '9.0'

  # Flutter.framework does not contain a i386 slice.
  s.libraries = ["c++", "z"]
  s.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES', 'EXCLUDED_ARCHS[sdk=iphonesimulator*]' => 'i386' }
  s.vendored_libraries = 'Frameworks/libpdfium.a'
  s.pod_target_xcconfig = { "OTHER_LDFLAGS" => "-force_load $(PODS_TARGET_SRCROOT)/Frameworks/libpdfium.a" }
  s.swift_version = '5.0'`

**Comment by jmgeffroy** [comment]
Hi @bitsydarel,

I am also struggling with the exact same issue, for PDFium too! Did you make any progress since your last message?
I'll post here if I find a solution.

Jean-Marie

**Comment by bitsydarel** [comment]
hi, @jmgeffroy 

Waiting on @dcharkes, because it’s seems to be a dart ffi issue as swift and obj can see and use the library.

**Comment by jmgeffroy** [comment]
Hi @bitsydarel,
Thanks a lot for your lightning-fast reply! OK, so I'll eagerly await, too ;-)

**Comment by dcharkes** [comment]
@bitsydarel I am able to look up symbols fine on iOS/MacOS with a small static library built by CMake using the podspec from https://github.com/dart-lang/sdk/issues/44328#issuecomment-855682903.

I presume either the static library is built in a different way, or your project setup is not the same. I haven't found the time to dig into your repo.

I'll see if I can upload my working repro somewhere so you can work from there.

**Comment by bitsydarel** [comment]
@dcharkes strangely we tried with other libraries that have dependency with other c/c++ libraries and combined the libraries using lipo.

Dart can’t see the symbols but rust and swift can, is there something different in how dart look for symbols for static libraries ?

**Comment by dcharkes** [comment]
Dart just uses `dlsym` on iOS and MacOS.

**Comment by jmgeffroy** [comment]
@bitsydarel Hi Darel, I don't know if you have fixed your issue, but today I spend a few minutes investigating, and I see that the problem is simply that the symbols exported by the static library become local in the generated Runner app.
In y case, objdump on the static library gives the following output:
```
objdump -t ./Frameworks/PDFium/libpdfium.a | fgrep InitLibrary
0000000000000000 g     F __TEXT,__text _FPDF_InitLibrary
000000000000000c g     F __TEXT,__text _FPDF_InitLibraryWithConfig
```
and objdump on the Runner app:
```
objdump -t /Users/hidden/Library/Developer/Xcode/DerivedData/Runner-ecfquixjrmwfygbvpmtorxrdxfth/Build/Products/Debug-iphoneos/Runner.app/Runner | fgrep InitLibrary
000000010001ac34 l     F __TEXT,__text _FPDF_InitLibrary
000000010001ac3c l     F __TEXT,__text _FPDF_InitLibraryWithConfig
```
As you can see the "g" has become "l". This is why you can call the PDFium functions from Swift but not dynamically load them from Dart (or Swift). 
These PDFium API function may have been compiled with the "hidden" visibility. The linkage of their symbols are thus changed from global to local.

There is a discussion about it on SO [here](https://stackoverflow.com/questions/54664759/keep-all-exported-symbols-when-creating-a-shared-library-from-a-static-library). The accepted answer provides a "unhide" script, but it should be adapted to your case and looks quite tricky. There is probably a better solution, such as rebuild libpdfium.a with -fvisibility=default.

I'll try to find some time to do some tests.

Hope this helps...

**Comment by zhengbomo** [comment]
any solution?

**Comment by orange-jacky** [comment]
I have the same issue, how to add Closed-source third-party library, such as libskrmc.a and skrmc.h by using ffigen for ios app
